nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractElasticsearchProcessor.java

```diff
@@ -0,0 +1,227 @@
+
+            final String hosts = context.getProperty(HOSTS).getValue();
+            esHosts = getEsHosts(hosts);
+
+            if (esHosts != null) {
+                for (final InetSocketAddress host : esHosts) {
+                    transportClient.addTransportAddress(new InetSocketTransportAddress(host));
+                }
+            }
+            esClient.set(transportClient);
+
+        } catch (Exception e) {
+            log.error("Failed to create Elasticsearch client due to {}", new Object[]{e}, e);
+            throw new ProcessException(e);
+        }
+    }
+
+    protected TransportClient getTransportClient(Settings.Builder settingsBuilder, String shieldUrl)
+            throws MalformedURLException {
+
+        // Create new transport client using the Builder pattern
+        TransportClient.Builder builder = TransportClient.builder();
+
+        // See if the Elasticsearch Shield JAR location was specified, and add the plugin if so
+
+        final ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
+        if (!StringUtils.isBlank(shieldUrl)) {
+            Thread.currentThread().setContextClassLoader(
+                    new URLClassLoader(new URL[]{new URL(shieldUrl)}, this.getClass().getClassLoader()));
+        }
+        try {
+            Class shieldPluginClass = Class.forName("org.elasticsearch.shield.ShieldPlugin", true,
+                    Thread.currentThread().getContextClassLoader());
+            builder = builder.addPlugin(shieldPluginClass);
+        } catch (ClassNotFoundException cnfe) {
+            getLogger().debug("Did not detect Elasticsearch Shield plugin, secure connections will not be available");
+        }
+
+        TransportClient transportClient = builder.settings(settingsBuilder.build()).build();
+        Thread.currentThread().setContextClassLoader(originalClassLoader);
+        return transportClient;
+    }
+
+    /**
+     * Dispose of ElasticSearch client
+     */
+    @OnStopped
+    public final void closeClient() {
+        if (esClient.get() != null) {
+            getLogger().info("Closing ElasticSearch Client");
+            esClient.get().close();
+            esClient.set(null);
+        }
+    }
+
+    /**
+     * Get the ElasticSearch hosts from a Nifi attribute, e.g.
+     *
+     * @param hosts A comma-separated list of ElasticSearch hosts (host:port,host2:port2, etc.)
+     * @return List of InetSocketAddresses for the ES hosts
+     */
+    private List<InetSocketAddress> getEsHosts(String hosts) {
+
+        if (hosts == null) {
+            return null;
+        }
+        final List<String> esList = Arrays.asList(hosts.split(","));
+        List<InetSocketAddress> esHosts = new ArrayList<>();
+
+        for (String item : esList) {
+
+            String[] addresses = item.split(":");
+            // Protect against invalid input like http://127.0.0.1:9300 (URL scheme should not be there)
```

Comment:
This should be included (implemented instead of just a todo comment). I tried PutElasticSearch with http://localhost:9300 and it blew up causing an out of memory error.